README.md
# MakeGallery

I like to show several pictures in some of my blog posts as a gallery
of sorts, with square thumbnails that link to the web-sized images.

I found a nifty one-line bash command that uses ImageMagick to
automate part of this process:

``` bash
#!/bin/bash

set -xv

: ${THUMBSIZE:=200}
: ${THUMBNAIL:=${THUMBSIZE}x${THUMBSIZE}}

mkdir -p thumbs
mogrify -format gif -path thumbs -thumbnail ${THUMBNAIL}^ -gravity center -extent ${THUMBNAIL} *.jpg
```

which creates small-ish .gif thumbnails in a `thumbs/` subdirectory.

In addition, I want to create web-scale .jpgs in a `web/` subdirectory
(mid-quality 1024 pixels wide).

## Installation

    $ gem install make_gallery

## Usage

**For up-to-the-minute usage information, you should run `make_gallery help`.**

    $ make_gallery (thumbs | web) [--source DIR] [--target DIR] [--size SIZE] [--format FORMAT] [--force] [--verbose] [--debug] [--dry-run]

* **thumbs:** create (square) thumbnails
* **web:** create web-sized images (maintains aspect ratio)
* *source:* the directory to take starting images from (default is current directory)
* *target:* the directory to write to (default is `./thumbs/` or `./web/`)
* *size:* horizontal size for web images, side size for thumbs
* *format:* JPG for web images, GIF for thumbs
* *force:* overwrite (default is abort if directory already exists)
* *dry-run:* don't actually create the images or directory, just show
  what would happen
* *verbose:* be chatty
* *debug:* be SUPER chatty about a lot of things


## Development

After checking out the repo, run `bin/setup` to install
dependencies. Then, run `bin/console` for an interactive prompt that
will allow you to experiment. Run `bundle exec make_gallery` to use
the code located in this directory, ignoring other installed copies of
this gem.

To install this gem onto your local machine, run `bundle exec rake
install`. To release a new version, update the version number in
`version.rb`, and then run `bundle exec rake release` to create a git
tag for the version, push git commits and tags, and push the `.gem`
file to [rubygems.org](https://rubygems.org).

## Contributing

Review the [Guidelines](CONTRIBUTING.md) for any information on how to contribute to this tool.

Make sure to also read the [Code of Conduct](CODE_OF_CONDUCT.md) if you wish to contribute.

1. Fork it ( https://github.com/tamouse/make_gallery/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

*"fork in, branch it, commit it, push it." -- daft punking our way through open source!*



CHANGELOG.md
# ChangeLog

# Next Release: 0.1.0

* First pre-release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'make_gallery'
require 'stringio'
require 'fileutils'

def save_output(&block)
  old_output = $stdout
  old_error = $stderr

  stro = StringIO.open('', 'w+')
  stre = StringIO.open('', 'w+')

  $stdout = stro
  $stderr = stre

  block_result = yield

  stro.rewind
  output = stro.read

  stre.rewind
  error = stre.read

  $stdout = old_output
  $stderr = old_error
  
  [ block_result, output, error ]
end


spec/lib/make_gallery_spec.rb
require 'spec_helper'
require 'make_gallery'

RSpec.describe MakeGallery::Thor do

  it "shows a banner" do
    result, output_buffer, error_buffer = save_output do 
      MakeGallery::Thor.start(%w[help])
    end
    expect(error_buffer).to be_empty, "Got errors: #{error_buffer}"
    expect(output_buffer).to match(/\ACommands:/), "Did not get a help banner. Output was: #{output_buffer}"
  end

  context "clean target directory" do
    let(:source) { File.expand_path("../../test_input_data/", __FILE__) }
    let(:target) { File.expand_path("../test_thumbs/", source) }
    let(:args)   do
      [
        "thumbs",
        "--source=#{source}",
        "--target=#{target}",
        "--dry-run"
      ]
    end

    it "creates thumbs" do
      FileUtils.rm_rf(target)
      result, output, error = save_output do
        MakeGallery::Thor.start(args)
      end
      expect(output).to include(%q|{"force"=>false, "verbose"=>true, "debug"=>true, "dry_run"=>true, "source"=>"/Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_input_data", "target"=>"/Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_thumbs", "size"=>200, "format"=>"gif", "quality"=>60}|)
      expect(output).to include(%q|Creating thumbs images in /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_input_data to /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_thumbs|)
      expect(output).to include(%q|/usr/local/bin/mogrify -verbose -format gif -path /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_thumbs -quality 60 -thumbnail 200x200^ -gravity center -extent 200x200 /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_input_data/squirrel-001.jpg /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_input_data/squirrel-002.jpg /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_input_data/squirrel-003.jpg /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_input_data/squirrel-004.jpg /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_input_data/squirrel-005.jpg /Users/tamara/Projects/rubystuff/scripts/make_gallery/spec/test_input_data/squirrel-006.jpg|)
      expect(output).to include(%q|Completed|)
      expect(error).to be_empty, "Error output not empty: #{error}"
    end
  end

  context "existing target directory" do
    let(:source) { File.expand_path("../../test_input_data/", __FILE__) }
    let(:target) { File.expand_path("../test_thumbs/", source) }
    let(:args)   do
      [
        "thumbs",
        "--source=#{source}",
        "--target=#{target}",
        "--dry-run"
      ]
    end

    it "gives a warning to use --force" do
      FileUtils.mkdir_p(target)
      result, output, error = save_output do
        MakeGallery::Thor.start(args)
      end
      expect(result).to match(%r|.*/test_thumbs exist. Use --force to overwrite|)

      # expect(1).to eq(0), "Result: #{result}\nOutput: #{output}\nError: #{error}"
    end

  end

end


spec/lib/make_gallery/command_spec.rb
require "spec_helper"
require "make_gallery/command"

RSpec.describe MakeGallery::Command do
  let(:source) { "./spec/test_input_data/" }
  let(:target) { "./spec/test_thumbs/" }
  let(:args) do
    {
      :debug => false,
      :dry_run => true,
      :verbose => true,
      :format => "gif",
      :quality => 30,
      :size => 200,
      :source => source,
      :target => target,
      :force => true
    }
  end
  subject { MakeGallery::Command.new(:thumbs, args) }

  describe "#build_command" do
    it "returns the mogrify command" do
      expect(subject.build_command).to match_array([
          "/usr/local/bin/mogrify",
          "-verbose",
          "-format gif",
          "-path ./spec/test_thumbs/",
          "-quality 30",
          "-thumbnail 200x200^",
          "-gravity center",
          "-extent 200x200",
          "./spec/test_input_data/squirrel-001.jpg",
          "./spec/test_input_data/squirrel-002.jpg",
          "./spec/test_input_data/squirrel-003.jpg",
          "./spec/test_input_data/squirrel-004.jpg",
          "./spec/test_input_data/squirrel-005.jpg",
          "./spec/test_input_data/squirrel-006.jpg"
        ])
    end
  end

  describe "#locate_mogrify" do
    context "when mogrify_path is NOT given" do
      context "when mogrify DOES exist" do
        it "returns the mogrify program location" do
          expect(File).to receive(:exist?).with("/usr/local/bin/mogrify").and_return(true)
          expect(subject.locate_mogrify).to eq("/usr/local/bin/mogrify")
        end
      end

      context "when mogrify does NOT exist" do
        it "prints an error and raises an exception" do
          expect(File).to receive(:exist?).with("/usr/local/bin/mogrify").and_return(false)
          expect{subject.locate_mogrify}.to raise_error
        end
      end
    end

    context "when mogrify_path IS given" do
      context "when mogrify DOES exist" do
        it "returns the mogrify program location" do
          subject.mogrify_path "/usr/local/bin/mogrify"
          expect(File).to receive(:exist?).with("/usr/local/bin/mogrify").and_return(true)
          expect(subject.locate_mogrify).to eq("/usr/local/bin/mogrify")
        end
      end

      context "when mogrify does NOT exist" do
        it "prints an error and raises an exception" do
          subject.mogrify_path "/usr/local/bin/mogrify"
          expect(File).to receive(:exist?).with("/usr/local/bin/mogrify").and_return(false)
          expect{subject.locate_mogrify}.to raise_error
        end
      end
    end
  end

  describe "#set_mogrify_options" do
    context "when the :thumbs action is used" do
      it "returns the correct command options" do
        expect(subject.set_mogrify_options).to match_array([
            "-verbose",
            "-format gif",
            "-path ./spec/test_thumbs/",
            "-quality 30",
            "-thumbnail 200x200^",
            "-gravity center",
            "-extent 200x200"
          ])
      end 
    end
    context "when the :web action is used" do
      it "returns the correct command options" do
        subject.action(:web)
        expect(subject.set_mogrify_options).to match_array([
            "-verbose",
            "-format gif",
            "-path ./spec/test_thumbs/",
            "-quality 30",
            "-resize 200"
          ])
      end 
    end
  end

  describe "#select_source_images" do
    it "returns the set of images" do
      expect(subject.select_source_images).to match_array([
          "./spec/test_input_data/squirrel-001.jpg",
          "./spec/test_input_data/squirrel-002.jpg",
          "./spec/test_input_data/squirrel-003.jpg",
          "./spec/test_input_data/squirrel-004.jpg",
          "./spec/test_input_data/squirrel-005.jpg",
          "./spec/test_input_data/squirrel-006.jpg"
        ])
    end 
  end
end


lib/make_gallery/command.rb
require 'fattr'
require 'highline'

module MakeGallery
  class Command 

    OUTLOG = HighLine.new($stdin, $stdout)
    ERRLOG = HighLine.new($stdin, $stderr)
    
    FATTRS = %w[action source target size format quality force verbose debug dry_run mogrify_path]

    fattrs *FATTRS

    def fattrs
      self.class.fattrs
    end


    def initialize(action, options)
      action(action)

      options.each do |k, v|
        public_send k, v
      end

      if debug
        fattrs.each do |attr|
          a = ERRLOG.color(attr, :cyan, :on_black)
          b = ERRLOG.color(public_send(attr).inspect, :white, :on_black)
          ERRLOG.say("DEBUG: #{a}: #{b}")
        end
      end

      fatal_say("#{target} exists. Use --force to overwrite") if File.exists?(target) && ! force
    end
    
    def to_h
      fattrs.inject({}) {|hash, attr| hash.update attr => public_send(attr) }
    end

    def process_images
      say("Creating #{action} images in #{source} to #{target}\n", :green)
      
      FileUtils.mkdir_p(target) unless dry_run

      cmd = build_command.join(" ")
      say("\nRunning: #{cmd}", :green)

      # skip the rest if it's a dry run
      unless dry_run
        o, e, s = Open3.capture3(cmd)
        fatal e unless s.success?
        say("\nOutput:\n", :green)
        say(o, :cyan)
        say(e, :cyan)
        say("\n")
      end

      say("\nCompleted\n", :green)
    end

    def build_command
      mogrify = locate_mogrify
      cmd_opts = set_mogrify_options
      source_images = select_source_images
      [mogrify, cmd_opts, source_images].flatten.tap do |t|
        debug_say("Command: #{t.inspect}")
      end
    end


    def locate_mogrify
      if mogrify_path
        fatal_say("Path given for mogrify command is wrong") unless File.exist?(mogrify_path)
        mogrify = mogrify_path
      else
        mogrify = %x(which mogrify).chomp
        fatal_say("Cannot find mogrify command. Use --mogrify_path option instead") unless File.exist?(mogrify)
      end
      debug_say("mogrify command: #{mogrify.inspect}")
      mogrify
    end

    def set_mogrify_options
      cmd_opts = []
      cmd_opts << "-verbose" if verbose
      cmd_opts << "-format #{format}"
      cmd_opts << "-path #{target}"
      cmd_opts << "-quality #{quality}"

      if action == :thumbs
        geometry = [size, size].join("x")
        cmd_opts << "-thumbnail #{geometry}^"
        cmd_opts << "-gravity center"
        cmd_opts << "-extent #{geometry}"
      elsif action == :web
        cmd_opts << "-resize #{size}"
      end

      debug_say "command opts: #{cmd_opts.inspect}"
      cmd_opts
    end

    def select_source_images
      debug_say "source: #{source}"
      debug_say "source files: #{Dir[File.join(source,'*')]}"
      source_images = Dir[File.join(source,'*')].grep(/jpe?g|png|gif/i)
      debug_say "source images: #{source_images.inspect}"
      fatal_say("No images found in #{source}!") if source_images.empty?
      source_images
    end

    def say(m, *colors)
      if colors.empty?
        colors = [:red, :on_black, :bold]
      end

      m = OUTLOG.color(m, *colors)
      OUTLOG.say(m) if verbose
    end      

    def fatal_say(m, *colors)
      if colors.empty?
        colors = [:red, :on_black, :bold]
      end

      m = ERRLOG.color(m, *colors)
      ERRLOG.say("FATAL: #{m}")

      raise MakeGallery::Exception.new(m)
    end

    def debug_say(m, *colors)
      if colors.empty?
        colors = [:yellow, :on_black, :bold]
      end

      m = ERRLOG.color(m, *colors)
      ERRLOG.say("DEBUG: #{m}") if debug
    end
  end
end


lib/make_gallery.rb
require "thor"
require "fileutils"
require "open3"
require "make_gallery/command"

module MakeGallery
  class Exception < RuntimeError; end
  
  class Thor < Thor

    include Thor::Actions

    class_option(:force,   type: :boolean, aliases: %w[-F], default: false,
      desc: "overwrite if directory or files exist")
    class_option(:verbose, type: :boolean, aliases: %w[-V], default: false,
      desc: "be chatty")
    class_option(:debug,   type: :boolean, aliases: %w[-D], default: false,
      desc: "be even *more* chatty")
    class_option(:dry_run, type: :boolean, aliases: %w[-n], default: false,
      desc: "Dry run, just show would happen, but do not do it")
    class_option(:mogrify_path, type: :string,
      desc: "path to mogrify command if make_gallery cannot find it")


    desc "thumbs", "Make small, square thumbnails of images in the source directory, saving them in the target directory."
    method_option(:source,  type: :string,  aliases: %w[-s], default: ".",
      desc: "source directory for images",
      banner: "DIR")
    method_option(:target,  type: :string,  aliases: %w[-t], default: "./thumbs/",
      desc: "destination directory for thumbnails",
      banner: "DIR")
    method_option(:size,    type: :numeric, aliases: %w[-w], default: 200,
      desc: "height and width in pixels")
    method_option(:format,  type: :string,  aliases: %w[-f], default: "gif",
      desc: "format of thumbnail")
    method_option(:quality, type: :numeric, aliases: %w[-q], default: 60,
      desc: "quality of image")
    def thumbs
      MakeGallery::Command.new(:thumbs, options).process_images
    rescue MakeGallery::Exception => e
      return e.to_s
    end

    desc "web", "Make web-sized images of the images in the source directory, saving them in the target directory. The new images retain the same aspect ratio as the source images."
    method_option(:source,  type: :string,  aliases: %w[-s], default: ".",
      desc: "source directory for images",
      banner: "DIR")
    method_option(:target,  type: :string,  aliases: %w[-t], default: "./web/",
      desc: "destination directory for thumbnails",
      banner: "DIR")
    method_option(:size,    type: :numeric, aliases: %w[-w], default: 1024,
      desc: "horizontal width in pixels")
    method_option(:format,  type: :string,  aliases: %w[-f], default: "jpg",
      desc: "format of thumbnail")
    method_option(:quality, type: :numeric, aliases: %w[-q], default: 60,
      desc: "quality of image")
    def web
      MakeGallery::Command.new(:web, options).process_images
    rescue MakeGallery::Exception => e
      return e.to_s
    end
    
    private

  end
end
